modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Assets.scala

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) 2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.fs2
+
+import java.net.URI
+import java.nio.file.{Path, Paths}
+
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+import cats.{Applicative, Parallel}
+import cats.implicits._
+
+import cats.effect.{Blocker, Concurrent, ConcurrentEffect, ContextShift, Resource, Sync, Timer}
+import cats.effect.concurrent.Ref
+
+import retry.{RetryDetails, RetryPolicies, RetryPolicy, retryingOnSomeErrors}
+
+import fs2.Stream
+import fs2.hash.md5
+import fs2.io.file.{copy, deleteIfExists, exists, readAll, tempFileResource, writeAll}
+
+import _root_.io.chrisdavenport.log4cats.Logger
+import _root_.io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+
+import com.snowplowanalytics.snowplow.enrich.fs2.io.Clients
+
+/**
+ * Functions responsible for periodic assets (such as MaxMind/IAB DBs) updates
+ * The common logic is to periodically invoke a function that:
+ * 1. Downloads a file (in background) to a temp location
+ * 2. Compares file's checksum with existing one (stored in a mutable hashmap)
+ * 3. If checksums match - delete the temp file, return
+ * 4. If checksums don't match - send a signal to stop raw stream
+ * (via `SignallingRef` in [[Environment]])
+ * 5. Once raw stream is stopped - delete an old file and move
+ * temp file to the old's file location
+ * If any of those URIs been updated and stopped the raw stream, it will be
+ * immediately resumed once the above procedure traversed all files
+ */
+object Assets {
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  /**
+   * State of the [[updateStream]], containing information about tracked URIs
+   * and `stop` signal from [[Environment]] as well as all clients necessary
+   * to download URIs
+   *
+   * @param files mutable hash map of URIs and their latest known state
+   * @param pauseEnrich stop signal coming from [[Environment]] and that can be used
+   *             to stop the raw stream consumption
+   * @param clients HTTP, GCS, S3 clients if necessary
+   */
+  final case class State[F[_]](
+    files: Ref[F, Map[URI, Hash]],
+    pauseEnrich: Ref[F, Boolean],
+    clients: Clients[F]
+  )
+
+  object State {
+
+    /** Test pair is used in tests to initialize HTTP client */
```

Comment:
Why do we have it here if it is used in tests?

Reply:
Because we need to skip it here https://github.com/snowplow/enrich/pull/328/files/72e8d30543915d21af2150223cb1598d043e754a#diff-638a4515052fa93eecdd441517d16652eb2cd047af5d77847997c83602d1eee9R92

Reply:
Looking at L92, it seems it isn't used in tests only, could you update this scaladoc?

Reply:
I didn't notice your comment and posted the last one, still scaladoc needs an update I think